File: backend/FitnessPlatform.Application/Domain/Documents/TrainingCompletion.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace FitnessPlatform.Application.Domain.Documents;

/// <summary>
/// MongoDB root aggregate tracking which exercises (and optionally sets) a client
/// has completed for a specific training session on a specific calendar date.
///
/// <para>
/// <b>Document shape chosen:</b> one document per (clientId, date, sessionId).
/// This is the cheapest shape for the two most common read paths:
/// (1) "did the client finish session X today?" — a single document lookup with a compound index;
/// (2) compliance roll-up — an indexed scan by (clientId, date) returns one document per session,
///     allowing the service to count completed sessions without pulling every exercise.
/// The alternative "one doc per (clientId, date)" with nested sessions would save one index entry
/// per day but would create wider documents and make fan-out writes (MarkWholeDayComplete) heavier
/// with multiple UpdateOne calls replaced by a single FindOneAndReplace, which isn't clearly cheaper.
/// </para>
/// </summary>
/// <remarks>
/// <b>Deprecated (#841).</b> Superseded by <see cref="SessionExecution"/>, which unifies this
/// document with <see cref="WorkoutLog"/>. The <c>trainingCompletions</c> collection is kept
/// read-only (no new writes) — do not add new write sites against this type. The one-shot
/// migration that would have folded these documents into <see cref="SessionExecution"/> was
/// deleted in #848 (there was no data to migrate), so nothing writes here at all any more.
/// Scheduled for removal in a follow-up chore.
/// </remarks>
public class TrainingCompletion
{
    /// <summary>
    /// MongoDB internal identifier.
    /// </summary>
    [BsonId]
    [BsonRepresentation(BsonType.ObjectId)]
    public ObjectId Id { get; set; }

    /// <summary>
    /// Public-facing identifier used in API requests and responses.
    /// </summary>
    [BsonElement("externalId")]
    public Guid ExternalId { get; set; }

    /// <summary>
    /// The client this completion record belongs to (matches ApplicationUser.Id — #840).
    /// </summary>
    [BsonElement("clientId")]
    public Guid ClientId { get; set; }

    /// <summary>
    /// The calendar date (midnight UTC) for which this completion applies.
    /// </summary>
    [BsonElement("date")]
    public DateTime Date { get; set; }

    /// <summary>
    /// The session (from the training plan) that was completed.
    /// Matches <see cref="TrainingSession.SessionId"/>.
    /// </summary>
    [BsonElement("sessionId")]
    public Guid SessionId { get; set; }

    /// <summary>
    /// Flat list of completed <see cref="SessionExercise.ExerciseId"/> instance values for this
    /// session on this date — both standalone exercises and exercises nested inside a workout.
    /// <para>
    /// Replaces the pre-#857-phase-3b <c>completedExerciseIdsBySection</c> dictionary (keyed by
    /// <see cref="TrainingWorkout.WorkoutId"/>, valued with catalog
    /// <see cref="SessionExercise.ExerciseExternalId"/>s), which could not distinguish two
    /// occurrences of the same catalog exercise within one workout or between a standalone
    /// occurrence and a nested one. See <see cref="SessionExecution.CompletedExerciseInstanceIds"/>
    /// for the live-model twin of this field.
    /// </para>
    /// </summary>
    [BsonElement("completedExerciseInstanceIds")]
    public List<Guid> CompletedExerciseInstanceIds { get; set; } = [];

    /// <summary>
    /// Workout IDs (matching <see cref="TrainingWorkout.WorkoutId"/>) that the
    /// client has marked complete on this date. Used for workouts that don't
    /// track at the exercise level — ForTime workouts that are just a name +
    /// time cap, e.g. "Running".
    /// </summary>
    [BsonElement("completedWorkoutIds")]
    [BsonIgnoreIfNull]
    public List<Guid>? CompletedWorkoutIds { get; set; }

    /// <summary>
    /// Optional per-set completion data, keyed by <see cref="SessionExercise.ExerciseExternalId"/>
    /// (serialized as a lowercase Guid string) — <b>NOT</b> rekeyed onto the per-instance
    /// <see cref="SessionExercise.ExerciseId"/> the way <see cref="CompletedExerciseInstanceIds"/>
    /// was. This document type is frozen/read-only (see class remarks) with no live write path,
    /// so no migration step exists (or ever ran) to resolve a catalog id against a specific plan
    /// session's instance ids here — see
    /// <see cref="SessionExecution.CompletedSets"/> for the full explanation and the matching
    /// reader in <c>GetFullTrainingPlanEndpoint</c>, which keys its lookup the same way.
    /// Each entry is the set of 1-based set numbers that were completed.
    /// Only populated when the client uses set-level tracking; absence means the
    /// exercise was marked complete at the exercise level only.
    /// </summary>
    [BsonElement("completedSets")]
    [BsonIgnoreIfNull]
    public Dictionary<string, List<int>>? CompletedSets { get; set; }

    /// <summary>
    /// When the first completion was recorded for this session-date combination.
    /// </summary>
    [BsonElement("dateCreated")]
    public DateTime DateCreated { get; set; }

    /// <summary>
    /// When this document was last updated.
    /// </summary>
    [BsonElement("dateUpdated")]
    [BsonIgnoreIfNull]
    public DateTime? DateUpdated { get; set; }

    /// <summary>
    /// Optimistic concurrency version. Incremented on each update.
    /// </summary>
    [BsonElement("version")]
    public int Version { get; set; } = 1;
}

File: mobile/src/api/trainingCompletion.ts
import api from './client';
import type {
  MarkExerciseCompleteRequest,
  MarkExerciseCompleteResponse,
  MarkExerciseIncompleteRequest,
  MarkExerciseIncompleteResponse,
  MarkWorkoutCompleteRequest,
  MarkWorkoutCompleteResponse,
  MarkWorkoutIncompleteRequest,
  MarkWorkoutIncompleteResponse,
  MarkSessionCompleteRequest,
  MarkSessionCompleteResponse,
  MarkSessionIncompleteRequest,
  MarkSessionIncompleteResponse,
  MarkWholeDayCompleteRequest,
  MarkWholeDayCompleteResponse,
  SessionCompletionSummary,
} from './generated';
import type { WodResult } from './wod-types';

// Re-export generated request/response types so consumer imports
// (`from '@/api/trainingCompletion'`) continue to work unchanged.
export type {
  MarkExerciseCompleteRequest,
  MarkExerciseCompleteResponse,
  MarkExerciseIncompleteRequest,
  MarkExerciseIncompleteResponse,
  MarkWorkoutCompleteRequest,
  MarkWorkoutCompleteResponse,
  MarkWorkoutIncompleteRequest,
  MarkWorkoutIncompleteResponse,
  MarkSessionCompleteRequest,
  MarkSessionCompleteResponse,
  MarkSessionIncompleteRequest,
  MarkSessionIncompleteResponse,
  MarkWholeDayCompleteRequest,
  MarkWholeDayCompleteResponse,
  SessionCompletionSummary,
};

// Re-export WodResult so callers can type-narrow their completion payloads.
export type { WodResult };

// ─── API calls ───────────────────────────────────────────────────────────────

/**
 * Mark a single exercise instance within a session as complete.
 * `exerciseId` is the per-instance id (SessionExercise.exerciseId), NOT the
 * catalog exerciseExternalId — the backend resolves the route segment
 * against the session's instance ids, so it distinguishes two placements of
 * the same catalog exercise in one session (nested + standalone, or nested
 * twice).
 * Idempotent — re-completing an already-complete exercise returns success.
 */
export async function markExerciseComplete(
  sessionId: string,
  exerciseId: string,
  request: MarkExerciseCompleteRequest,
): Promise<MarkExerciseCompleteResponse> {
  const { data } = await api.post<MarkExerciseCompleteResponse>(
    `/client/training/sessions/${sessionId}/exercises/${exerciseId}/complete`,
    request,
  );
  return data;
}

/**
 * Remove the completion mark for a single exercise instance within a session.
 * `exerciseId` is the per-instance id, leaving other placements of the same
 * catalog exercise in the same session untouched.
 * Idempotent — if the exercise is already not marked complete, returns success.
 */
export async function markExerciseIncomplete(
  sessionId: string,
  exerciseId: string,
  request: MarkExerciseIncompleteRequest,
): Promise<MarkExerciseIncompleteResponse> {
  const { data } = await api.delete<MarkExerciseIncompleteResponse>(
    `/client/training/sessions/${sessionId}/exercises/${exerciseId}/complete`,
    { data: request },
  );
  return data;
}

/**
 * Mark a single workout within a session as complete.
 * Used for workouts that don't track at the exercise level — typically
 * ForTime workouts that are just a name + time cap (e.g. "Running").
 * Idempotent.
 */
export async function markWorkoutComplete(
  sessionId: string,
  workoutId: string,
  request: MarkWorkoutCompleteRequest = {},
): Promise<MarkWorkoutCompleteResponse> {
  const { data } = await api.post<MarkWorkoutCompleteResponse>(
    `/client/training/sessions/${sessionId}/workouts/${workoutId}/complete`,
    request,
  );
  return data;
}

/**
 * Remove the completion mark for a single workout within a session.
 * Idempotent.
 */
export async function markWorkoutIncomplete(
  sessionId: string,
  workoutId: string,
  request: MarkWorkoutIncompleteRequest = {},
): Promise<MarkWorkoutIncompleteResponse> {
  const { data } = await api.delete<MarkWorkoutIncompleteResponse>(
    `/client/training/sessions/${sessionId}/workouts/${workoutId}/complete`,
    { data: request },
  );
  return data;
}

/**
 * Mark an entire training session as complete (fans out to all exercises).
 * Idempotent.
 */
export async function markSessionComplete(
  sessionId: string,
  request: MarkSessionCompleteRequest = {},
): Promise<MarkSessionCompleteResponse> {
  const { data } = await api.post<MarkSessionCompleteResponse>(
    `/client/training/sessions/${sessionId}/complete`,
    request,
  );
  return data;
}

/**
 * Remove the completion mark for an entire training session.
 * Idempotent.
 */
export async function markSessionIncomplete(
  sessionId: string,
  request: MarkSessionIncompleteRequest = {},
): Promise<MarkSessionIncompleteResponse> {
  const { data } = await api.delete<MarkSessionIncompleteResponse>(
    `/client/training/sessions/${sessionId}/complete`,
    { data: request },
  );
  return data;
}

/**
 * Mark every training session scheduled for a calendar day complete.
 * Resolves sessions from the active plan's week/day-of-week mapping.
 * Idempotent — already-complete sessions are skipped.
 */
export async function markWholeDayComplete(
  request: MarkWholeDayCompleteRequest = {},
): Promise<MarkWholeDayCompleteResponse> {
  const { data } = await api.post<MarkWholeDayCompleteResponse>(
    '/client/training/day/complete',
    request,
  );
  return data;
}
